
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Pipelines to turn basic genomic data into Ensembl cores and back.">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/ensembl/io/genomio/seq_region/collection/">
      
      
        <link rel="prev" href="../">
      
      
        <link rel="next" href="../dump/">
      
      
      <link rel="icon" href="../../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>collection - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/stylesheets/main.e53b48f4.min.css">
      
        
        <link rel="stylesheet" href="../../../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.seq_region.collection" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              collection
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of Conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    ensembl
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    brc4
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1" id="__nav_5_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/runnable/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    runnable
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1_1" id="__nav_5_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_fasta
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_report
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    fill_metadata
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_schema_factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load_sequence_data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare_genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    seqregion_parser
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1_2" id="__nav_5_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    io
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genomio
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1" id="__nav_5_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../annotation/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    annotation
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_1" id="__nav_5_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            annotation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../annotation/update_description/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    update_description
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_2" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../assembly/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    assembly
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_2" id="__nav_5_1_2_1_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_2">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/status/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    status
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_3" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../data/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    data
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_3" id="__nav_5_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/external_db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    external_db_map
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/gff3/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/schemas/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_4" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../database/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    database
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_4" id="__nav_5_1_2_1_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_4">
            <span class="md-nav__icon md-icon"></span>
            database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/core_server/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    core_server
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/dbconnection_lite/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dbconnection_lite
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/meta_getter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    meta_getter
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_5" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../events/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    events
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_5" id="__nav_5_1_2_1_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_5">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_6" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../external_db/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    external_db
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_6" id="__nav_5_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            external_db
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../external_db/db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    db_map
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_7" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../fasta/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    fasta
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_7" id="__nav_5_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            fasta
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/chunk/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    chunk
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_8" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genbank/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genbank
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_8" id="__nav_5_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/extract_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_9" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_metadata/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_metadata
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_9" id="__nav_5_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/extend/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extend
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_10" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_stats/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_stats
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_10" id="__nav_5_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            genome_stats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_11" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gff3/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_11" id="__nav_5_1_2_1_11_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_11_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_11">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/extract_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_annotation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/features/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    features
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/gene_merger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gene_merger
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/id_allocator/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    id_allocator
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/overlaps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    overlaps
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/restructure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    restructure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/simplifier/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    simplifier
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_12" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../manifest/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_12" id="__nav_5_1_2_1_12_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_12_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_12">
            <span class="md-nav__icon md-icon"></span>
            manifest
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/check_integrity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    check_integrity
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/compute_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compute_stats
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/generate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    generate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/manifest/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/manifest_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest_stats
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13" id="__nav_5_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/json/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    json
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13_1" id="__nav_5_1_2_1_13_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_5_1_2_1_13_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13_1">
            <span class="md-nav__icon md-icon"></span>
            json
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/validate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_14" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    seq_region
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_14" id="__nav_5_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_14_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            seq_region
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    collection
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    collection
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;collection
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" collection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqRegionDict" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegionDict
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;SeqCollection
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" SeqCollection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.mock" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;mock
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.seqs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;seqs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_mitochondrial_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_mitochondrial_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_translation_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_gbff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_gbff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_report" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seq_region_from_report" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;make_seq_region_from_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seqregion_from_gbff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;make_seqregion_from_gbff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.remove" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.to_list" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_list
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gbff/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gbff
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mappings/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    mappings
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    report
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_15" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    utils
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_15" id="__nav_5_1_2_1_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_15">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/json_utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running Nextflow
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../pipelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;collection
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" collection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqRegionDict" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegionDict
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;SeqCollection
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" SeqCollection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.mock" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;mock
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.seqs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;seqs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_mitochondrial_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_mitochondrial_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_translation_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_gbff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_gbff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_report" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seq_region_from_report" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;make_seq_region_from_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seqregion_from_gbff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;make_seqregion_from_gbff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.remove" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.to_list" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_list
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>collection</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.seq_region.collection" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.io.genomio.seq_region.collection</code>


<a href="#ensembl.io.genomio.seq_region.collection" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>SeqCollection as a collection of seq_regions metadata.</p>










  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.collection.SeqRegionDict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.collection.SeqRegionDict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>



<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.seq_region.collection.SeqCollection" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>SeqCollection</code>


<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>Represent a collection of seq_regions metadata.</p>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">SeqCollection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Represent a collection of seq_regions metadata.&quot;&quot;&quot;</span>

    <span class="n">mock</span><span class="p">:</span> <span class="nb">bool</span>
    <span class="n">seqs</span><span class="p">:</span> <span class="nb">dict</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mock</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mock</span> <span class="o">=</span> <span class="n">mock</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">from_gbff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gbff_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Store seq_regions extracted from a GBFF file.</span>

<span class="sd">        If a seq_region with the same ID exists in the collection, it will be replaced.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
                <span class="n">record_data</span> <span class="o">=</span> <span class="n">GBFFRecord</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
                <span class="n">new_seq</span><span class="p">:</span> <span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_seqregion_from_gbff</span><span class="p">(</span><span class="n">record_data</span><span class="p">)</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
                <span class="n">merged_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge</span><span class="p">(</span><span class="n">new_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{}))</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_seq</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">:</span> <span class="n">SeqRegionDict</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">SeqRegionDict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Merge a source dict in a destination dict (only add values or append to lists).&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">destination</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">source</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">source</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                <span class="n">destination</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="n">value</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">destination</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

        <span class="k">return</span> <span class="n">destination</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">make_seqregion_from_gbff</span><span class="p">(</span><span class="n">record_data</span><span class="p">:</span> <span class="n">GBFFRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a seq_region dict extracted from a GBFF record.&quot;&quot;&quot;</span>
        <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">record_data</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)}</span>  <span class="c1"># type: ignore[arg-type]</span>

        <span class="k">if</span> <span class="n">record_data</span><span class="o">.</span><span class="n">is_circular</span><span class="p">():</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Is there a genetic code defined?</span>
        <span class="n">codon_table</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_codon_table</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">codon_table</span>

        <span class="c1"># Is it an organelle?</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_organelle</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location</span>

        <span class="c1"># Is there a comment stating the Genbank record this is based on?</span>
        <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_genbank_id</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">genbank_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">genbank_id</span><span class="p">}]</span>

        <span class="k">return</span> <span class="n">seqr</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">from_report</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">report_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Store seq_regions extracted from an INSDC assembly report file.</span>

<span class="sd">        If a seq_region with the same id exists in the collection, it will be replaced.</span>

<span class="sd">        Args:</span>
<span class="sd">            report_path: Path to the sequence regions report file.</span>
<span class="sd">            is_refseq: True if the source of the report is RefSeq, false if INSDC.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">report</span> <span class="o">=</span> <span class="n">ReportRecord</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seq_data</span> <span class="ow">in</span> <span class="n">report</span><span class="o">.</span><span class="n">reader</span><span class="p">:</span>
            <span class="n">new_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_seq_region_from_report</span><span class="p">(</span><span class="n">seq_data</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">new_seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">merged_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge</span><span class="p">(</span><span class="n">new_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{}))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_seq</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">make_seq_region_from_report</span><span class="p">(</span>
        <span class="n">seq_data</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">synonym_map</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SYNONYM_MAP</span><span class="p">,</span>
        <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a sequence region from the information provided.</span>

<span class="sd">        An empty sequence region will be returned if no accession information is found.</span>

<span class="sd">        Args:</span>
<span class="sd">            data: Dict from the report representing one line, where the key is the column name.</span>
<span class="sd">            is_refseq: True if the source is RefSeq, false if INSDC.</span>
<span class="sd">            synonym_map: Map of INSDC report column names to sequence region field names.</span>
<span class="sd">            molecule_location: Map of sequence type to SO location.</span>

<span class="sd">        Raises:</span>
<span class="sd">            UnknownMetadata: If the sequence role or location is not recognised.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Set accession as the sequence region name</span>
        <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span> <span class="k">if</span> <span class="n">is_refseq</span> <span class="k">else</span> <span class="s2">&quot;GenBank&quot;</span>
        <span class="n">accession_id</span> <span class="o">=</span> <span class="n">seq_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s2">-Accn&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">accession_id</span> <span class="ow">or</span> <span class="p">(</span><span class="n">accession_id</span> <span class="o">==</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No </span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s1"> accession ID found for </span><span class="si">{</span><span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Sequence-Name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{}</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession_id</span>

        <span class="c1"># Add synonyms</span>
        <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">synonym_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">seq_data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
                <span class="n">synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]}</span>
                <span class="n">synonyms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">synonym</span><span class="p">)</span>
        <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">])</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>

        <span class="c1"># Add sequence length</span>
        <span class="n">field</span> <span class="o">=</span> <span class="s2">&quot;Sequence-Length&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">seq_data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">])</span>

        <span class="c1"># Add coordinate system and location</span>
        <span class="n">seq_role</span> <span class="o">=</span> <span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Sequence-Role&quot;</span><span class="p">]</span>
        <span class="c1"># Scaffold?</span>
        <span class="k">if</span> <span class="n">seq_role</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;unplaced-scaffold&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized-scaffold&quot;</span><span class="p">):</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scaffold&quot;</span>
        <span class="c1"># Chromosome? Check location</span>
        <span class="k">elif</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s2">&quot;assembled-molecule&quot;</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Assigned-Molecule-Location/Type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="c1"># Get location metadata</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">exc</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence role: </span><span class="si">{</span><span class="n">seq_role</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">seq_region</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">remove</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove seq_regions based on a provided list of accessions.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">:</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot exclude seq not found: </span><span class="si">{</span><span class="n">seq_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">add_translation_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">location_codon</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">LOCATION_CODON</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the translation codon table to each sequence region (when missing) based on its location.</span>

<span class="sd">        Args:</span>
<span class="sd">            location_codon: Map of known codon tables for known locations.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">location_codon</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location_codon</span><span class="p">[</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">add_mitochondrial_codon_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</span>

<span class="sd">        If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</span>

<span class="sd">        Args:</span>
<span class="sd">            taxon_id: The species taxon ID.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mock</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Skip mitochondrial codon table: mock&quot;</span><span class="p">)</span>
            <span class="k">return</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">taxon_id</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Skip mitochondrial codon table: no taxon_id to use&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/taxonomy/rest/tax-id/</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/json&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
        <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
        <span class="c1"># In case we have been redirected, check for HTML opening tag</span>
        <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Response from </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2"> is not JSON&quot;</span><span class="p">)</span>
        <span class="n">decoded</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
        <span class="n">genetic_code</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">decoded</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mitochondrialGeneticCode&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">genetic_code</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No mitochondria genetic code found for taxon </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_code</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">to_list</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the sequences as a simple list of `SeqRegionDict` objects.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.mock" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">mock</span> <span class="o">=</span> <span class="n">mock</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.mock" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.seqs" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">seqs</span> <span class="o">=</span> <span class="p">{}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.seqs" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.add_mitochondrial_codon_table" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_mitochondrial_codon_table" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</p>
<p>If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The species taxon ID.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_mitochondrial_codon_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</span>

<span class="sd">    If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id: The species taxon ID.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mock</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Skip mitochondrial codon table: mock&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">taxon_id</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Skip mitochondrial codon table: no taxon_id to use&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/taxonomy/rest/tax-id/</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/json&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
    <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
    <span class="c1"># In case we have been redirected, check for HTML opening tag</span>
    <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Response from </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2"> is not JSON&quot;</span><span class="p">)</span>
    <span class="n">decoded</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="n">genetic_code</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">decoded</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mitochondrialGeneticCode&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">genetic_code</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No mitochondria genetic code found for taxon </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_code</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.add_translation_table" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">add_translation_table</span><span class="p">(</span><span class="n">location_codon</span><span class="o">=</span><span class="n">LOCATION_CODON</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.add_translation_table" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Adds the translation codon table to each sequence region (when missing) based on its location.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>location_codon</code>
            </td>
            <td>
                  <code><span title="typing.Mapping">Mapping</span>[<span title="str">str</span>, <span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Map of known codon tables for known locations.</p>
              </div>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-attribute&quot;&gt;&lt;/code&gt;            &lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;LOCATION_CODON&lt;/span&gt; &lt;span class=&quot;o&quot;&gt;=&lt;/span&gt; &lt;span class=&quot;n&quot;&gt;MappingProxyType&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;({&lt;/span&gt;&lt;span class=&quot;s1&quot;&gt;&amp;#39;apicoplast_chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;mi&quot;&gt;4&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;})&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-module-attribute&quot;&gt;&lt;code&gt;module-attribute&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt; (&lt;code&gt;ensembl.io.genomio.seq_region.mappings.LOCATION_CODON&lt;/code&gt;)" href="../mappings/#ensembl.io.genomio.seq_region.mappings.LOCATION_CODON">LOCATION_CODON</a></code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_translation_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">location_codon</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">LOCATION_CODON</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the translation codon table to each sequence region (when missing) based on its location.</span>

<span class="sd">    Args:</span>
<span class="sd">        location_codon: Map of known codon tables for known locations.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">location_codon</span><span class="p">:</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location_codon</span><span class="p">[</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.from_gbff" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">from_gbff</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_gbff" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Store seq_regions extracted from a GBFF file.</p>
<p>If a seq_region with the same ID exists in the collection, it will be replaced.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">from_gbff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gbff_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Store seq_regions extracted from a GBFF file.</span>

<span class="sd">    If a seq_region with the same ID exists in the collection, it will be replaced.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="n">record_data</span> <span class="o">=</span> <span class="n">GBFFRecord</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="n">new_seq</span><span class="p">:</span> <span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_seqregion_from_gbff</span><span class="p">(</span><span class="n">record_data</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
            <span class="n">merged_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge</span><span class="p">(</span><span class="n">new_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{}))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_seq</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.from_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">from_report</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="n">is_refseq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.from_report" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Store seq_regions extracted from an INSDC assembly report file.</p>
<p>If a seq_region with the same id exists in the collection, it will be replaced.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>report_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the sequence regions report file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>is_refseq</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the source of the report is RefSeq, false if INSDC.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">from_report</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">report_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Store seq_regions extracted from an INSDC assembly report file.</span>

<span class="sd">    If a seq_region with the same id exists in the collection, it will be replaced.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the sequence regions report file.</span>
<span class="sd">        is_refseq: True if the source of the report is RefSeq, false if INSDC.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">report</span> <span class="o">=</span> <span class="n">ReportRecord</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">seq_data</span> <span class="ow">in</span> <span class="n">report</span><span class="o">.</span><span class="n">reader</span><span class="p">:</span>
        <span class="n">new_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_seq_region_from_report</span><span class="p">(</span><span class="n">seq_data</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">new_seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="n">merged_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge</span><span class="p">(</span><span class="n">new_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_seq</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.make_seq_region_from_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">make_seq_region_from_report</span><span class="p">(</span><span class="n">seq_data</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">,</span> <span class="n">synonym_map</span><span class="o">=</span><span class="n">SYNONYM_MAP</span><span class="p">,</span> <span class="n">molecule_location</span><span class="o">=</span><span class="n">MOLECULE_LOCATION</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seq_region_from_report" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a sequence region from the information provided.</p>
<p>An empty sequence region will be returned if no accession information is found.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>data</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict from the report representing one line, where the key is the column name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>is_refseq</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the source is RefSeq, false if INSDC.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>synonym_map</code>
            </td>
            <td>
                  <code><span title="typing.Mapping">Mapping</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Map of INSDC report column names to sequence region field names.</p>
              </div>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-attribute&quot;&gt;&lt;/code&gt;            &lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;SYNONYM_MAP&lt;/span&gt; &lt;span class=&quot;o&quot;&gt;=&lt;/span&gt; &lt;span class=&quot;n&quot;&gt;MappingProxyType&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;({&lt;/span&gt;&lt;span class=&quot;s1&quot;&gt;&amp;#39;Assigned-Molecule&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;INSDC&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;GenBank-Accn&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;GenBank&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;RefSeq-Accn&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;RefSeq&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;Sequence-Name&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;INSDC_submitted_name&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;})&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-module-attribute&quot;&gt;&lt;code&gt;module-attribute&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt; (&lt;code&gt;ensembl.io.genomio.seq_region.mappings.SYNONYM_MAP&lt;/code&gt;)" href="../mappings/#ensembl.io.genomio.seq_region.mappings.SYNONYM_MAP">SYNONYM_MAP</a></code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>molecule_location</code>
            </td>
            <td>
                  <code><span title="typing.Mapping">Mapping</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Map of sequence type to SO location.</p>
              </div>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-attribute&quot;&gt;&lt;/code&gt;            &lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;MOLECULE_LOCATION&lt;/span&gt; &lt;span class=&quot;o&quot;&gt;=&lt;/span&gt; &lt;span class=&quot;n&quot;&gt;MappingProxyType&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;({&lt;/span&gt;&lt;span class=&quot;s1&quot;&gt;&amp;#39;apicoplast&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;apicoplast_chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;nuclear_chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;kinetoplast&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;kinetoplast_chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;linkage group&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;linkage_group&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;mitochondrion&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;mitochondrial_chromosome&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;,&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;plasmid&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;:&lt;/span&gt; &lt;span class=&quot;s1&quot;&gt;&amp;#39;plasmid&amp;#39;&lt;/span&gt;&lt;span class=&quot;p&quot;&gt;})&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-module-attribute&quot;&gt;&lt;code&gt;module-attribute&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt; (&lt;code&gt;ensembl.io.genomio.seq_region.mappings.MOLECULE_LOCATION&lt;/code&gt;)" href="../mappings/#ensembl.io.genomio.seq_region.mappings.MOLECULE_LOCATION">MOLECULE_LOCATION</a></code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;UnknownMetadata&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.seq_region.exceptions.UnknownMetadata&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.seq_region.exceptions.UnknownMetadata">UnknownMetadata</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the sequence role or location is not recognised.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">make_seq_region_from_report</span><span class="p">(</span>
    <span class="n">seq_data</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">synonym_map</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SYNONYM_MAP</span><span class="p">,</span>
    <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a sequence region from the information provided.</span>

<span class="sd">    An empty sequence region will be returned if no accession information is found.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: Dict from the report representing one line, where the key is the column name.</span>
<span class="sd">        is_refseq: True if the source is RefSeq, false if INSDC.</span>
<span class="sd">        synonym_map: Map of INSDC report column names to sequence region field names.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        UnknownMetadata: If the sequence role or location is not recognised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Set accession as the sequence region name</span>
    <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span> <span class="k">if</span> <span class="n">is_refseq</span> <span class="k">else</span> <span class="s2">&quot;GenBank&quot;</span>
    <span class="n">accession_id</span> <span class="o">=</span> <span class="n">seq_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s2">-Accn&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">accession_id</span> <span class="ow">or</span> <span class="p">(</span><span class="n">accession_id</span> <span class="o">==</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No </span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s1"> accession ID found for </span><span class="si">{</span><span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Sequence-Name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
    <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession_id</span>

    <span class="c1"># Add synonyms</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">synonym_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">seq_data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]}</span>
            <span class="n">synonyms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">synonym</span><span class="p">)</span>
    <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">])</span>
    <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>

    <span class="c1"># Add sequence length</span>
    <span class="n">field</span> <span class="o">=</span> <span class="s2">&quot;Sequence-Length&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">seq_data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq_data</span><span class="p">[</span><span class="n">field</span><span class="p">])</span>

    <span class="c1"># Add coordinate system and location</span>
    <span class="n">seq_role</span> <span class="o">=</span> <span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Sequence-Role&quot;</span><span class="p">]</span>
    <span class="c1"># Scaffold?</span>
    <span class="k">if</span> <span class="n">seq_role</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;unplaced-scaffold&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized-scaffold&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scaffold&quot;</span>
    <span class="c1"># Chromosome? Check location</span>
    <span class="k">elif</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s2">&quot;assembled-molecule&quot;</span><span class="p">:</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">seq_data</span><span class="p">[</span><span class="s2">&quot;Assigned-Molecule-Location/Type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="c1"># Get location metadata</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">exc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence role: </span><span class="si">{</span><span class="n">seq_role</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">seq_region</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.make_seqregion_from_gbff" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">make_seqregion_from_gbff</span><span class="p">(</span><span class="n">record_data</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.make_seqregion_from_gbff" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a seq_region dict extracted from a GBFF record.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">make_seqregion_from_gbff</span><span class="p">(</span><span class="n">record_data</span><span class="p">:</span> <span class="n">GBFFRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a seq_region dict extracted from a GBFF record.&quot;&quot;&quot;</span>
    <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">record_data</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)}</span>  <span class="c1"># type: ignore[arg-type]</span>

    <span class="k">if</span> <span class="n">record_data</span><span class="o">.</span><span class="n">is_circular</span><span class="p">():</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="c1"># Is there a genetic code defined?</span>
    <span class="n">codon_table</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_codon_table</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">codon_table</span>

    <span class="c1"># Is it an organelle?</span>
    <span class="n">location</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_organelle</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location</span>

    <span class="c1"># Is there a comment stating the Genbank record this is based on?</span>
    <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">record_data</span><span class="o">.</span><span class="n">get_genbank_id</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">genbank_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">genbank_id</span><span class="p">}]</span>

    <span class="k">return</span> <span class="n">seqr</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.remove" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">remove</span><span class="p">(</span><span class="n">to_exclude</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.remove" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Remove seq_regions based on a provided list of accessions.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">remove</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove seq_regions based on a provided list of accessions.&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">:</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot exclude seq not found: </span><span class="si">{</span><span class="n">seq_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.seq_region.collection.SeqCollection.to_list" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">to_list</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.seq_region.collection.SeqCollection.to_list" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns the sequences as a simple list of <code>SeqRegionDict</code> objects.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/collection.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">to_list</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequences as a simple list of `SeqRegionDict` objects.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../../../../..", "features": ["content.tooltips", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../../../../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>