
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Pipelines to turn basic genomic data into Ensembl cores and back.">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/ensembl/io/genomio/gff3/simplifier/">
      
      
        <link rel="prev" href="../restructure/">
      
      
        <link rel="next" href="../../manifest/">
      
      
      <link rel="icon" href="../../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>simplifier - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/stylesheets/main.e53b48f4.min.css">
      
        
        <link rel="stylesheet" href="../../../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.gff3.simplifier" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              simplifier
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of Conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    ensembl
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    brc4
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1" id="__nav_5_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/runnable/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    runnable
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1_1" id="__nav_5_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_fasta
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_report
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    fill_metadata
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_schema_factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load_sequence_data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare_genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    seqregion_parser
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1_2" id="__nav_5_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    io
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genomio
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1" id="__nav_5_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../annotation/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    annotation
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_1" id="__nav_5_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            annotation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../annotation/update_description/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    update_description
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_2" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../assembly/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    assembly
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_2" id="__nav_5_1_2_1_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_2">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/status/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    status
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_3" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../data/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    data
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_3" id="__nav_5_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/external_db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    external_db_map
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/gff3/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/schemas/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_4" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../database/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    database
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_4" id="__nav_5_1_2_1_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_4">
            <span class="md-nav__icon md-icon"></span>
            database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/core_server/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    core_server
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/dbconnection_lite/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dbconnection_lite
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../database/meta_getter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    meta_getter
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_5" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../events/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    events
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_5" id="__nav_5_1_2_1_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_5">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../events/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_6" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../external_db/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    external_db
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_6" id="__nav_5_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            external_db
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../external_db/db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    db_map
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_7" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../fasta/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    fasta
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_7" id="__nav_5_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            fasta
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/chunk/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    chunk
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_8" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genbank/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genbank
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_8" id="__nav_5_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/extract_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_9" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_metadata/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_metadata
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_9" id="__nav_5_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/extend/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extend
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_10" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_stats/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_stats
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_10" id="__nav_5_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            genome_stats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_11" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_11" id="__nav_5_1_2_1_11_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_11_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1_11">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../extract_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_annotation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../features/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    features
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_merger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gene_merger
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../id_allocator/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    id_allocator
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../overlaps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    overlaps
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../restructure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    restructure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    simplifier
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    simplifier
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;simplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" simplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFSimplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.allow_pseudogene_with_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;allow_pseudogene_with_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.annotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;annotations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.exclude_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;exclude_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.fail_types" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;fail_types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.genome" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;genome
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;records
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.refseq" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;refseq
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.skip_unrecognized" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;skip_unrecognized
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.stable_ids" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;stable_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.clean_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_gene_for_lone_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_transcript" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_gene_for_lone_transcript
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.format_gene_segments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_gene_segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.get_provider_name" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_provider_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_mirna" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_mirna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_non_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_non_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_pseudogene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_pseudogene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_transcripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3_feature
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Records
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records.from_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_gff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records.to_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_gff
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_12" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../manifest/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_12" id="__nav_5_1_2_1_12_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_12_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_12">
            <span class="md-nav__icon md-icon"></span>
            manifest
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/check_integrity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    check_integrity
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/compute_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compute_stats
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/generate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    generate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/manifest/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/manifest_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest_stats
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13" id="__nav_5_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/json/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    json
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13_1" id="__nav_5_1_2_1_13_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_5_1_2_1_13_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13_1">
            <span class="md-nav__icon md-icon"></span>
            json
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/validate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_14" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../seq_region/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    seq_region
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_14" id="__nav_5_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            seq_region
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/collection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    collection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/gbff/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gbff
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/mappings/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    mappings
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    report
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_15" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    utils
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_15" id="__nav_5_1_2_1_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_15">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/json_utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running Nextflow
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../pipelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;simplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" simplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFSimplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.allow_pseudogene_with_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;allow_pseudogene_with_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.annotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;annotations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.exclude_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;exclude_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.fail_types" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;fail_types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.genome" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;genome
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;records
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.refseq" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;refseq
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.skip_unrecognized" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;skip_unrecognized
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.stable_ids" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;stable_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.clean_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_gene_for_lone_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_transcript" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_gene_for_lone_transcript
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.format_gene_segments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_gene_segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.get_provider_name" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_provider_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_mirna" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_mirna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_non_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_non_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_pseudogene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_pseudogene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_transcripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3_feature
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Records
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records.from_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;from_gff
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.simplifier.Records.to_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_gff
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>simplifier</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.gff3.simplifier" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.io.genomio.gff3.simplifier</code>


<a href="#ensembl.io.genomio.gff3.simplifier" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Standardize the gene model representation of a GFF3 file, and extract the functional annotation
in a separate file.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>GFFSimplifier</code>


<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</p>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFParserError&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.exceptions.GFFParserError&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.gff3.exceptions.GFFParserError">GFFParserError</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If an error cannot be automatically fixed.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">GFFSimplifier</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</span>

<span class="sd">    Raises:</span>
<span class="sd">        GFFParserError: If an error cannot be automatically fixed.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genome_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">skip_unrecognized</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">allow_pseudogene_with_cds</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create an object that simplifies `SeqFeature` objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            genome_path: Genome metadata file.</span>
<span class="sd">            skip_unrecognized: Do not include unknown biotypes instead of raising an exception.</span>
<span class="sd">            allow_pseudogene_with_cds: Keep CDSs under pseudogenes that have them. Delete them otherwise.</span>

<span class="sd">        Raises:</span>
<span class="sd">            GFFParserError: If a biotype is unknown and `skip_unrecognized` is False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="n">skip_unrecognized</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_cds</span> <span class="o">=</span> <span class="n">allow_pseudogene_with_cds</span>

        <span class="c1"># Load biotypes</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">files</span><span class="p">(</span><span class="n">ensembl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">genomio</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">gff3</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="s2">&quot;biotypes.json&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">as_file</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="k">as</span> <span class="n">biotypes_json</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">biotypes_json</span><span class="p">)</span>

        <span class="c1"># Load genome metadata</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">genome_path</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">genome_fh</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">genome_fh</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">refseq</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">refseq</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Other preparations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span> <span class="o">=</span> <span class="n">StableIDAllocator</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">set_prefix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">:</span> <span class="n">Set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="c1"># Init the actual data we will store</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">Records</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">FunctionalAnnotations</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_provider_name</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_provider_name</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the provider name for this genome.</span>

<span class="sd">        If this information is not available, will try to infer it from the assembly accession. Will</span>
<span class="sd">        return &quot;GenBank&quot; otherwise.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">provider_name</span> <span class="o">=</span> <span class="s2">&quot;GenBank&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">provider_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF&quot;</span><span class="p">):</span>
                    <span class="n">provider_name</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No genome file, using the default provider_name: </span><span class="si">{</span><span class="n">provider_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">provider_name</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a</span>
<span class="sd">        functional annotation file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">from_gff</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
            <span class="n">cleaned_features</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="n">split_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_mirna</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">split_genes</span><span class="p">:</span>
                    <span class="n">cleaned_features</span> <span class="o">+=</span> <span class="n">split_genes</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">clean_feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
                        <span class="n">cleaned_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_feature</span><span class="p">)</span>
                    <span class="k">except</span> <span class="p">(</span><span class="n">UnsupportedFeatureError</span><span class="p">,</span> <span class="n">IgnoredFeatureError</span><span class="p">)</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">err</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
            <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">cleaned_features</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">:</span>
            <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">))</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="s2">&quot;Unrecognized types found, abort&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">simpler_gff3_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a simpler version of a GFF3 feature.</span>

<span class="sd">        Raises:</span>
<span class="sd">            IgnoredFeatureError: If the feature type is ignored.</span>
<span class="sd">            UnsupportedFeatureError: If the feature type is not supported.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Special cases</span>
        <span class="n">non_gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_non_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">non_gene</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">non_gene</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="n">IgnoredFeatureError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ignored type </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Synonym</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein_coding_gene&quot;</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>

        <span class="c1"># Lone sub-gene features, create a gene</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_gene_for_lone_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_gene_for_lone_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># What to do with unsupported gene types</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;gene=</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">UnsupportedFeatureError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported type </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Normalize and store</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">store_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_gene_for_lone_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA/mRNA, and &#39;ncRNA_gene&#39; for all others.</span>

<span class="sd">        Args:</span>
<span class="sd">            feat: The transcript for which we want to create a gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">feat</span>

        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">):</span>
            <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
        <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">feat</span><span class="p">]</span>

        <span class="c1"># Use the transcript ID for the gene, and generate a sub ID for the transcript</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">new_gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

        <span class="c1"># Remove the exon/CDS parent so it is properly updated</span>
        <span class="k">for</span> <span class="n">subfeat</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Parent&quot;</span><span class="p">]</span>

        <span class="c1"># Check if it&#39;s a pseudogene</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
            <span class="n">is_pseudo</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">subfeat</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="n">pseudo_qual</span> <span class="o">=</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pseudo&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">and</span> <span class="n">pseudo_qual</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span>
                    <span class="n">is_pseudo</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">del</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">]</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="n">is_pseudo</span><span class="p">:</span>
                <span class="n">new_gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>

        <span class="k">return</span> <span class="n">new_gene</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_gene_for_lone_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.</span>

<span class="sd">        Args:</span>
<span class="sd">            feat: The CDS for which we want to create a gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">feat</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Create a transcript, add the CDS</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">feat</span><span class="p">]</span>

        <span class="c1"># Add an exon too</span>
        <span class="n">exon</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
        <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

        <span class="c1"># Create a gene, add the transcript</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
            <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
            <span class="k">del</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">]</span>
        <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_gene_id</span><span class="p">()</span>
        <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">new_gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span>

        <span class="k">return</span> <span class="n">new_gene</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_non_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalised &quot;non-gene&quot; or `None` if not applicable.</span>

<span class="sd">        Only transposable elements supported at the moment.</span>

<span class="sd">        Args:</span>
<span class="sd">            feat: Feature to normalise.</span>

<span class="sd">        Raises:</span>
<span class="sd">            NotImplementedError: If the feature is a not supported non-gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;non_gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
            <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
            <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_mobile_genetic_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
            <span class="c1"># Generate ID if needed</span>
            <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">refseq</span><span class="p">)</span>
            <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="c1"># This is a failsafe in case you add supported non-genes</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported non-gene: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_normalize_mobile_genetic_element</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalize a mobile element if it has a mobile_element_type field.&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">mobile_element_type</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;mobile_element_type&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No &#39;mobile_element_type&#39; tag found&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>

        <span class="c1"># Get the type (and name) from the attrib</span>
        <span class="n">element_type</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">partition</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
        <span class="n">description</span> <span class="o">=</span> <span class="n">element_type</span>
        <span class="k">if</span> <span class="n">element_name</span><span class="p">:</span>
            <span class="n">description</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">)&quot;</span>

        <span class="c1"># Keep the metadata in the description if the type is known</span>
        <span class="k">if</span> <span class="n">element_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;transposon&quot;</span><span class="p">,</span> <span class="s2">&quot;retrotransposon&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">):</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">description</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">feat</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;mobile_element_type&#39; is not a transposon: </span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">clean_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the same gene without qualifiers unrelated to the gene structure.&quot;&quot;&quot;</span>

        <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>
        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="c1"># Replace qualifiers</span>
            <span class="n">old_transcript_qualifiers</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_transcript_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
            <span class="p">}</span>

            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="n">old_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
                <span class="p">}</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: Gene object to normalize.</span>
<span class="sd">            functional_annotation: List of feature annotations (appended by this method).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">refseq</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">refseq</span><span class="p">)</span>
        <span class="n">restructure_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalize CDSs if allowed, otherwise remove them.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
            <span class="k">return</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_cds</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_pseudogene_cds_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalizes a transcript.&quot;&quot;&quot;</span>

        <span class="n">allowed_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
        <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>

        <span class="n">transcripts_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_transcript_types</span>
                <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ignored_transcript_types</span>
            <span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;transcript=</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized transcript type: </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="sa">f</span><span class="s2">&quot; for </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="n">transcripts_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># New transcript ID</span>
            <span class="n">transcript_number</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript_number</span><span class="p">)</span>

            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

            <span class="c1"># EXONS AND CDS</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcript_subfeatures</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">transcript</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">transcripts_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">transcripts_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">        Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript: Gene segment transcript feature.</span>

<span class="sd">        Raises:</span>
<span class="sd">            GeneSegmentError: Unable to get the segment type information from the feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">transcript</span>

        <span class="c1"># Guess the segment type from the transcript attribs</span>
        <span class="n">seg_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_segment_type</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">seg_type</span><span class="p">:</span>
            <span class="c1"># Get the information from a CDS instead</span>
            <span class="n">sub_feats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span>
            <span class="n">cdss</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="n">sub_feats</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">cdss</span><span class="p">:</span>
                <span class="n">seg_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_segment_type</span><span class="p">(</span><span class="n">cdss</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">seg_type</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GeneSegmentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unable to infer segment from </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seg_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_segment&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_get_segment_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Infer if a segment is &quot;IG&quot; (immunoglobulin) of &quot;TR&quot; (t-cell) from the feature attribs.</span>

<span class="sd">        Returns an empty string if no segment type info was found.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">product</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;standard_name&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">product</span><span class="p">:</span>
            <span class="n">product</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">product</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>

        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\b(immunoglobulin|ig)\b&quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;IG&quot;</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\bt[- _]cell\b&quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;TR&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_normalize_transcript_subfeatures</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">GFFSeqFeature</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript with normalized sub-features.&quot;&quot;&quot;</span>
        <span class="n">exons_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">exon_number</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">tcount</span><span class="p">,</span> <span class="n">feat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="c1"># New exon ID</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">-E</span><span class="si">{</span><span class="n">exon_number</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">exon_number</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Replace qualifiers</span>
                <span class="n">old_exon_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_exon_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="c1"># New CDS ID</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]:</span>
                    <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sub_transcript=</span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized exon type for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (for transcript </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                <span class="k">continue</span>

        <span class="k">if</span> <span class="n">exons_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">exons_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_mirna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns gene representations from a miRNA gene that can be loaded in an Ensembl database.</span>

<span class="sd">        Change the representation from the form `gene[ primary_transcript[ exon, miRNA[ exon ] ] ]`</span>
<span class="sd">        to `ncRNA_gene[ miRNA_primary_transcript[ exon ] ]` and `gene[ miRNA[ exon ] ]`</span>

<span class="sd">        Raises:</span>
<span class="sd">            GFFParserError: If gene has more than 1 transcript, the transcript was not formatted</span>
<span class="sd">                correctly or there are unknown sub-features.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">base_id</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span>
        <span class="n">transcripts</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span>

        <span class="c1"># Insert main gene first if needed</span>
        <span class="n">old_gene</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;primary_transcript&quot;</span><span class="p">:</span>
            <span class="n">primary</span> <span class="o">=</span> <span class="n">old_gene</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">primary</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">primary</span><span class="p">]</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="n">primary</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">transcripts</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_id</span><span class="si">}</span><span class="s2">_0&quot;</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span>

        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">transcripts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;primary_transcript&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene has too many sub_features for miRNA </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Passed the checks</span>
        <span class="n">primary</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">primary</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;miRNA_primary_transcript&quot;</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Formatting miRNA gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">new_genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_primary_subfeatures</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">num</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">sub</span> <span class="ow">in</span> <span class="n">primary</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sub</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="n">new_primary_subfeatures</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">sub</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;miRNA&quot;</span><span class="p">:</span>
                <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">num</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">sub</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">new_gene_id</span><span class="p">)</span>
                <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">sub</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">new_gene_id</span><span class="p">}</span>
                <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">sub</span><span class="p">]</span>
                <span class="n">new_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown subtypes for miRNA features: </span><span class="si">{</span><span class="n">sub</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">primary</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_primary_subfeatures</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">new_genes</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Primary_transcript without miRNA in </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">all_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">all_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">+</span> <span class="n">new_genes</span>

        <span class="c1"># Normalize like other genes</span>
        <span class="n">all_genes_cleaned</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">new_gene</span> <span class="ow">in</span> <span class="n">all_genes</span><span class="p">:</span>
            <span class="n">new_gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">store_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
            <span class="n">all_genes_cleaned</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">all_genes_cleaned</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.allow_pseudogene_with_cds" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">allow_pseudogene_with_cds</span> <span class="o">=</span> <span class="n">allow_pseudogene_with_cds</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.allow_pseudogene_with_cds" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.annotations" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">annotations</span> <span class="o">=</span> <span class="n">FunctionalAnnotations</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_provider_name</span><span class="p">())</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.annotations" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.exclude_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">exclude_seq_regions</span> <span class="o">=</span> <span class="p">[]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.exclude_seq_regions" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.fail_types" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">fail_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.fail_types" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.genome" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">genome</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">genome_fh</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.genome" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.records" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">records</span> <span class="o">=</span> <span class="n">Records</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.records" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.refseq" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">refseq</span> <span class="o">=</span> <span class="kc">False</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.refseq" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.skip_unrecognized" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="n">skip_unrecognized</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.skip_unrecognized" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.stable_ids" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">stable_ids</span> <span class="o">=</span> <span class="n">StableIDAllocator</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.stable_ids" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.clean_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.clean_gene" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Return the same gene without qualifiers unrelated to the gene structure.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">clean_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the same gene without qualifiers unrelated to the gene structure.&quot;&quot;&quot;</span>

    <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>
    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="c1"># Replace qualifiers</span>
        <span class="n">old_transcript_qualifiers</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_transcript_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="n">old_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_cds" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">create_gene_for_lone_cds</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_cds" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a gene created for a lone CDS.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>feat</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFSeqFeature&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.features.GFFSeqFeature&lt;/code&gt;)" href="../features/#ensembl.io.genomio.gff3.features.GFFSeqFeature">GFFSeqFeature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The CDS for which we want to create a gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_gene_for_lone_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.</span>

<span class="sd">    Args:</span>
<span class="sd">        feat: The CDS for which we want to create a gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">feat</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Create a transcript, add the CDS</span>
    <span class="n">transcript</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">feat</span><span class="p">]</span>

    <span class="c1"># Add an exon too</span>
    <span class="n">exon</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
    <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

    <span class="c1"># Create a gene, add the transcript</span>
    <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
        <span class="k">del</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">]</span>
    <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_gene_id</span><span class="p">()</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">new_gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span>

    <span class="k">return</span> <span class="n">new_gene</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_transcript" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">create_gene_for_lone_transcript</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.create_gene_for_lone_transcript" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a gene for lone transcripts: 'gene' for tRNA/rRNA/mRNA, and 'ncRNA_gene' for all others.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>feat</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFSeqFeature&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.features.GFFSeqFeature&lt;/code&gt;)" href="../features/#ensembl.io.genomio.gff3.features.GFFSeqFeature">GFFSeqFeature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The transcript for which we want to create a gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_gene_for_lone_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA/mRNA, and &#39;ncRNA_gene&#39; for all others.</span>

<span class="sd">    Args:</span>
<span class="sd">        feat: The transcript for which we want to create a gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">feat</span>

    <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">):</span>
        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
    <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">feat</span><span class="p">]</span>

    <span class="c1"># Use the transcript ID for the gene, and generate a sub ID for the transcript</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
    <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_gene</span><span class="o">.</span><span class="n">id</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">new_gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

    <span class="c1"># Remove the exon/CDS parent so it is properly updated</span>
    <span class="k">for</span> <span class="n">subfeat</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Parent&quot;</span><span class="p">]</span>

    <span class="c1"># Check if it&#39;s a pseudogene</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
        <span class="n">is_pseudo</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">subfeat</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="n">pseudo_qual</span> <span class="o">=</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pseudo&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">and</span> <span class="n">pseudo_qual</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span>
                <span class="n">is_pseudo</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">del</span> <span class="n">subfeat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">]</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">is_pseudo</span><span class="p">:</span>
            <span class="n">new_gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>

    <span class="k">return</span> <span class="n">new_gene</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.format_gene_segments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.format_gene_segments" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns the equivalent Ensembl biotype feature for gene segment transcript features.</p>
<p>Supported features: "C_gene_segment" and "V_gene_segment".</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcript</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFSeqFeature&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.features.GFFSeqFeature&lt;/code&gt;)" href="../features/#ensembl.io.genomio.gff3.features.GFFSeqFeature">GFFSeqFeature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene segment transcript feature.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GeneSegmentError&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.exceptions.GeneSegmentError&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.gff3.exceptions.GeneSegmentError">GeneSegmentError</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Unable to get the segment type information from the feature.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">    Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript: Gene segment transcript feature.</span>

<span class="sd">    Raises:</span>
<span class="sd">        GeneSegmentError: Unable to get the segment type information from the feature.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="c1"># Guess the segment type from the transcript attribs</span>
    <span class="n">seg_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_segment_type</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">seg_type</span><span class="p">:</span>
        <span class="c1"># Get the information from a CDS instead</span>
        <span class="n">sub_feats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span>
        <span class="n">cdss</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="n">sub_feats</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">cdss</span><span class="p">:</span>
            <span class="n">seg_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_segment_type</span><span class="p">(</span><span class="n">cdss</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">seg_type</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GeneSegmentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unable to infer segment from </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seg_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_segment&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.get_provider_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">get_provider_name</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.get_provider_name" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns the provider name for this genome.</p>
<p>If this information is not available, will try to infer it from the assembly accession. Will
return "GenBank" otherwise.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_provider_name</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the provider name for this genome.</span>

<span class="sd">    If this information is not available, will try to infer it from the assembly accession. Will</span>
<span class="sd">    return &quot;GenBank&quot; otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">provider_name</span> <span class="o">=</span> <span class="s2">&quot;GenBank&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">provider_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF&quot;</span><span class="p">):</span>
                <span class="n">provider_name</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No genome file, using the default provider_name: </span><span class="si">{</span><span class="n">provider_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">provider_name</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_gene" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a normalized gene structure, separate from the functional elements.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFSeqFeature&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.features.GFFSeqFeature&lt;/code&gt;)" href="../features/#ensembl.io.genomio.gff3.features.GFFSeqFeature">GFFSeqFeature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene object to normalize.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>functional_annotation</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of feature annotations (appended by this method).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene object to normalize.</span>
<span class="sd">        functional_annotation: List of feature annotations (appended by this method).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">refseq</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">refseq</span><span class="p">)</span>
    <span class="n">restructure_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_mirna" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">normalize_mirna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_mirna" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns gene representations from a miRNA gene that can be loaded in an Ensembl database.</p>
<p>Change the representation from the form <code>gene[ primary_transcript[ exon, miRNA[ exon ] ] ]</code>
to <code>ncRNA_gene[ miRNA_primary_transcript[ exon ] ]</code> and <code>gene[ miRNA[ exon ] ]</code></p>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFParserError&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.exceptions.GFFParserError&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.gff3.exceptions.GFFParserError">GFFParserError</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If gene has more than 1 transcript, the transcript was not formatted
correctly or there are unknown sub-features.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normalize_mirna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns gene representations from a miRNA gene that can be loaded in an Ensembl database.</span>

<span class="sd">    Change the representation from the form `gene[ primary_transcript[ exon, miRNA[ exon ] ] ]`</span>
<span class="sd">    to `ncRNA_gene[ miRNA_primary_transcript[ exon ] ]` and `gene[ miRNA[ exon ] ]`</span>

<span class="sd">    Raises:</span>
<span class="sd">        GFFParserError: If gene has more than 1 transcript, the transcript was not formatted</span>
<span class="sd">            correctly or there are unknown sub-features.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">base_id</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span>
    <span class="n">transcripts</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span>

    <span class="c1"># Insert main gene first if needed</span>
    <span class="n">old_gene</span> <span class="o">=</span> <span class="n">gene</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;primary_transcript&quot;</span><span class="p">:</span>
        <span class="n">primary</span> <span class="o">=</span> <span class="n">old_gene</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">primary</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">primary</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="n">primary</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">transcripts</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_id</span><span class="si">}</span><span class="s2">_0&quot;</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span>

    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">transcripts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;primary_transcript&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene has too many sub_features for miRNA </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Passed the checks</span>
    <span class="n">primary</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">primary</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;miRNA_primary_transcript&quot;</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Formatting miRNA gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">new_genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">new_primary_subfeatures</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">num</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">for</span> <span class="n">sub</span> <span class="ow">in</span> <span class="n">primary</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">sub</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="n">new_primary_subfeatures</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">sub</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;miRNA&quot;</span><span class="p">:</span>
            <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">num</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GFFSeqFeature</span><span class="p">(</span><span class="n">sub</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">new_gene_id</span><span class="p">)</span>
            <span class="n">new_gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">sub</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">new_gene_id</span><span class="p">}</span>
            <span class="n">new_gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">sub</span><span class="p">]</span>
            <span class="n">new_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown subtypes for miRNA features: </span><span class="si">{</span><span class="n">sub</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">primary</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_primary_subfeatures</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">new_genes</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Primary_transcript without miRNA in </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">all_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">all_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">+</span> <span class="n">new_genes</span>

    <span class="c1"># Normalize like other genes</span>
    <span class="n">all_genes_cleaned</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">new_gene</span> <span class="ow">in</span> <span class="n">all_genes</span><span class="p">:</span>
        <span class="n">new_gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">store_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
        <span class="n">all_genes_cleaned</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">new_gene</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">all_genes_cleaned</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_non_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">normalize_non_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_non_gene" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Returns a normalised "non-gene" or <code>None</code> if not applicable.</p>
<p>Only transposable elements supported at the moment.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>feat</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;GFFSeqFeature&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.features.GFFSeqFeature&lt;/code&gt;)" href="../features/#ensembl.io.genomio.gff3.features.GFFSeqFeature">GFFSeqFeature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Feature to normalise.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="NotImplementedError">NotImplementedError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the feature is a not supported non-gene.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normalize_non_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">GFFSeqFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a normalised &quot;non-gene&quot; or `None` if not applicable.</span>

<span class="sd">    Only transposable elements supported at the moment.</span>

<span class="sd">    Args:</span>
<span class="sd">        feat: Feature to normalise.</span>

<span class="sd">    Raises:</span>
<span class="sd">        NotImplementedError: If the feature is a not supported non-gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;non_gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
        <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_mobile_genetic_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="c1"># Generate ID if needed</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">refseq</span><span class="p">)</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
    <span class="c1"># This is a failsafe in case you add supported non-genes</span>
    <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported non-gene: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_pseudogene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">normalize_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_pseudogene" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Normalize CDSs if allowed, otherwise remove them.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normalize_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Normalize CDSs if allowed, otherwise remove them.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
        <span class="k">return</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_cds</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">normalize_pseudogene_cds_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_transcripts" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.normalize_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Normalizes a transcript.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normalize_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Normalizes a transcript.&quot;&quot;&quot;</span>

    <span class="n">allowed_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
    <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>

    <span class="n">transcripts_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_transcript_types</span>
            <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ignored_transcript_types</span>
        <span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;transcript=</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unrecognized transcript type: </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="sa">f</span><span class="s2">&quot; for </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="p">)</span>
            <span class="n">transcripts_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># New transcript ID</span>
        <span class="n">transcript_number</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_ids</span><span class="o">.</span><span class="n">generate_transcript_id</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript_number</span><span class="p">)</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

        <span class="c1"># EXONS AND CDS</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcript_subfeatures</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">transcript</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">transcripts_to_delete</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">transcripts_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">simpler_gff3</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a
functional annotation file.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a</span>
<span class="sd">    functional annotation file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">from_gff</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">:</span>
        <span class="n">cleaned_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="n">split_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_mirna</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">split_genes</span><span class="p">:</span>
                <span class="n">cleaned_features</span> <span class="o">+=</span> <span class="n">split_genes</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">clean_feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
                    <span class="n">cleaned_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_feature</span><span class="p">)</span>
                <span class="k">except</span> <span class="p">(</span><span class="n">UnsupportedFeatureError</span><span class="p">,</span> <span class="n">IgnoredFeatureError</span><span class="p">)</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">err</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">cleaned_features</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">:</span>
        <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="s2">&quot;Unrecognized types found, abort&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3_feature" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.GFFSimplifier.simpler_gff3_feature" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Creates a simpler version of a GFF3 feature.</p>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;IgnoredFeatureError&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.exceptions.IgnoredFeatureError&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.gff3.exceptions.IgnoredFeatureError">IgnoredFeatureError</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the feature type is ignored.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;UnsupportedFeatureError&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.gff3.exceptions.UnsupportedFeatureError&lt;/code&gt;)" href="../exceptions/#ensembl.io.genomio.gff3.exceptions.UnsupportedFeatureError">UnsupportedFeatureError</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the feature type is not supported.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">simpler_gff3_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">GFFSeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GFFSeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates a simpler version of a GFF3 feature.</span>

<span class="sd">    Raises:</span>
<span class="sd">        IgnoredFeatureError: If the feature type is ignored.</span>
<span class="sd">        UnsupportedFeatureError: If the feature type is not supported.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Special cases</span>
    <span class="n">non_gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_non_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">non_gene</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">non_gene</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="n">IgnoredFeatureError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ignored type </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Synonym</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein_coding_gene&quot;</span><span class="p">:</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>

    <span class="c1"># Lone sub-gene features, create a gene</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_gene_for_lone_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_gene_for_lone_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># What to do with unsupported gene types</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;gene=</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">UnsupportedFeatureError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported type </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Normalize and store</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">store_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.simplifier.Records" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>Records</code>


<a href="#ensembl.io.genomio.gff3.simplifier.Records" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="list">list</span></code></p>


        <p>List of GFF3 SeqRecords.</p>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Records</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List of GFF3 SeqRecords.&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">from_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">excluded</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Loads records from a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            in_gff_path: Input GFF3 file path.</span>
<span class="sd">            excluded: Record IDs to not load from the GFF3 file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">excluded</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">excluded</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">excluded</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> - in exclusion list&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">clean_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="n">clean_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_record</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Writes the current list of records in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
            <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.Records.from_gff" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">from_gff</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="n">excluded</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.Records.from_gff" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Loads records from a GFF3 file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>in_gff_path</code>
            </td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input GFF3 file path.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>excluded</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[<span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Record IDs to not load from the GFF3 file.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">from_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">excluded</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Loads records from a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        in_gff_path: Input GFF3 file path.</span>
<span class="sd">        excluded: Record IDs to not load from the GFF3 file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">excluded</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">excluded</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">excluded</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> - in exclusion list&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">clean_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="n">clean_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_record</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.gff3.simplifier.Records.to_gff" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">to_gff</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.simplifier.Records.to_gff" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Writes the current list of records in a GFF3 file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>out_gff_path</code>
            </td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to GFF3 file where to write the records.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Writes the current list of records in a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
        <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../../../../..", "features": ["content.tooltips", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../../../../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>