
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Pipelines to turn basic genomic data into Ensembl cores and back.">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/ensembl/io/genomio/fasta/">
      
      
        <link rel="prev" href="../external_db/db_map/">
      
      
        <link rel="next" href="chunk/">
      
      
      <link rel="icon" href="../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>fasta - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/stylesheets/main.e53b48f4.min.css">
      
        
        <link rel="stylesheet" href="../../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.fasta" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              fasta
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../../../img/ensembl-logo.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl/ensembl-genomio
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Code of Conduct
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    ensembl
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../brc4/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    brc4
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1" id="__nav_5_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../brc4/runnable/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    runnable
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_1_1" id="__nav_5_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_fasta
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare_report
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    fill_metadata
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_schema_factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load_sequence_data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare_genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    seqregion_parser
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1_2" id="__nav_5_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    io
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genomio
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1" id="__nav_5_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_5_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../annotation/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    annotation
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_1" id="__nav_5_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            annotation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/update_description/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    update_description
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_2" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../assembly/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    assembly
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_2" id="__nav_5_1_2_1_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_2">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/status/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    status
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_3" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../data/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    data
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_3" id="__nav_5_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/external_db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    external_db_map
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/gff3/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/schemas/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_4" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../database/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    database
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_4" id="__nav_5_1_2_1_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_4">
            <span class="md-nav__icon md-icon"></span>
            database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/core_server/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    core_server
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/dbconnection_lite/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dbconnection_lite
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/meta_getter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    meta_getter
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_5" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../events/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    events
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_5" id="__nav_5_1_2_1_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_5">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_6" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../external_db/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    external_db
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_6" id="__nav_5_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            external_db
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../external_db/db_map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    db_map
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_7" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="./" class="md-nav__link md-nav__link--active">
              
  
  
  <span class="md-ellipsis">
    fasta
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link md-nav__link--active" for="__nav_5_1_2_1_7" id="__nav_5_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            fasta
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="chunk/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    chunk
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_8" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../genbank/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genbank
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_8" id="__nav_5_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/download/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    download
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/extract_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_9" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../genome_metadata/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_metadata
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_9" id="__nav_5_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/extend/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extend
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_10" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../genome_stats/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    genome_stats
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_10" id="__nav_5_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            genome_stats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/compare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_11" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../gff3/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    gff3
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_11" id="__nav_5_1_2_1_11_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_11_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_11">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/extract_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract_annotation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/features/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    features
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/gene_merger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gene_merger
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/id_allocator/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    id_allocator
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/overlaps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    overlaps
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/process/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/restructure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    restructure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/simplifier/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    simplifier
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_12" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../manifest/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_12" id="__nav_5_1_2_1_12_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_12_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_12">
            <span class="md-nav__icon md-icon"></span>
            manifest
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/check_integrity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    check_integrity
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/compute_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    compute_stats
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/generate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    generate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/manifest/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/manifest_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    manifest_stats
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    schemas
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13" id="__nav_5_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_13_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/json/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    json
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_13_1" id="__nav_5_1_2_1_13_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_5_1_2_1_13_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_13_1">
            <span class="md-nav__icon md-icon"></span>
            json
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/factory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    factory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/validate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_14" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../seq_region/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    seq_region
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_14" id="__nav_5_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            seq_region
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/collection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    collection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/dump/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dump
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/exceptions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    exceptions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/gbff/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    gbff
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/mappings/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    mappings
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/prepare/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    prepare
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    report
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1_2_1_15" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../utils/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    utils
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1_2_1_15" id="__nav_5_1_2_1_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_5_1_2_1_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1_2_1_15">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/json_utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    json_utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running Nextflow
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../pipelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;fasta
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" fasta">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;CompareFasta
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" CompareFasta">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.comp" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;comp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.fasta_core" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;fasta_core
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.fasta_ext" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;fasta_ext
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.output_dir" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;output_dir
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.build_seq_dict" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;build_seq_dict
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.compare_seq_for_Ns" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;compare_seq_for_Ns
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.compare_seqs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;compare_seqs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.find_common_groups" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;find_common_groups
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.read_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;read_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.CompareFasta.write_results" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;write_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.FastaParserError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;FastaParserError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.SeqGroup" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;SeqGroup
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" SeqGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.SeqGroup.count" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;count
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.SeqGroup.ids" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.SeqGroup.add_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_id
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.check_chunk_size_and_tolerance" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_chunk_size_and_tolerance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.chunk_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;chunk_fasta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.chunk_fasta_stream" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;chunk_fasta_stream
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.get_peptides_to_exclude" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_peptides_to_exclude
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.get_tolerated_size" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_tolerated_size
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.prep_fasta_data" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prep_fasta_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.prepare_out_dir_for_individuals" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prepare_out_dir_for_individuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.split_seq_by_chunk_size" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;split_seq_by_chunk_size
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.fasta.split_seq_by_n" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;split_seq_by_n
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>fasta</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.fasta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.io.genomio.fasta</code>


<a href="#ensembl.io.genomio.fasta" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Fasta files processing module.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.fasta.CompareFasta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>CompareFasta</code>


<a href="#ensembl.io.genomio.fasta.CompareFasta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>Read and compare the FASTA sequences.</p>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">CompareFasta</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read and compare the FASTA sequences.&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_ext</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">fasta_core</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize the `CompareFasta` with input fasta files and output directory.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_ext: Path to INSDC fasta file.</span>
<span class="sd">            fasta_core: Path to core db fasta file.</span>
<span class="sd">            output_dir: Directory where comparison logs will be stored.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta_ext</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_ext</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta_core</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_core</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">compare_seqs</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare two FASTA files for common, unique, and differing sequences.</span>
<span class="sd">        Use `write_results()` to generate a report.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq_ext</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_ext</span><span class="p">)</span>
        <span class="n">seq_core</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_core</span><span class="p">)</span>

        <span class="c1"># Compare sequences</span>
        <span class="n">seq_ext_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_seq_dict</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span>
        <span class="n">seq_core_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_seq_dict</span><span class="p">(</span><span class="n">seq_core</span><span class="p">)</span>

        <span class="c1"># Compare number of sequences</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_core</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="s2">&quot;WARNING: Different number of sequences: &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;fasta_ext [ n = </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span><span class="si">}</span><span class="s2"> ] -Vs- fasta_core [ n = </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_core</span><span class="p">)</span><span class="si">}</span><span class="s2"> ]&quot;</span>
            <span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Different number of sequences: fasta_ext compared to fasta_core&quot;</span><span class="p">)</span>

        <span class="n">common</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">find_common_groups</span><span class="p">(</span><span class="n">seq_ext_dict</span><span class="p">,</span> <span class="n">seq_core_dict</span><span class="p">)</span>

        <span class="c1"># Sequences that are not common</span>
        <span class="n">only1</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq</span><span class="p">:</span> <span class="n">group</span> <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="p">}</span>

        <span class="n">only2</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq</span><span class="p">:</span> <span class="n">group</span> <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="p">}</span>

        <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequences only in Fasta_1: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only1</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequences only in Fasta_2: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only2</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Common ids: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">common_ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">common_ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">common</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Check for sequences with extra Ns</span>
        <span class="k">if</span> <span class="n">only1</span> <span class="ow">and</span> <span class="n">only2</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">compare_seq_for_Ns</span><span class="p">(</span><span class="n">only1</span><span class="p">,</span> <span class="n">only2</span><span class="p">)</span>

        <span class="c1"># Write results to file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">write_results</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Reads a FASTA file and returns a dictionary mapping sequence IDs to sequences.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_path: Path to the FASTA file. Supports gzipped files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dictionary where keys are sequence IDs and values are sequences with all non-CGTA</span>
<span class="sd">                characters replaced by &quot;N&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read fasta file </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span>
                <span class="n">sequences</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[^CGTA]&quot;</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
        <span class="k">return</span> <span class="n">sequences</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">build_seq_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seqs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Builds a dictionary of unique sequences and their associated IDs, accounting for duplicates.</span>

<span class="sd">        Args:</span>
<span class="sd">            seqs: A dictionary where keys are sequence IDs and values are sequences.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dictionary where keys are unique sequences and values are `SeqGroup` objects that</span>
<span class="sd">                group sequence IDs sharing the same sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seqs_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs_dict</span><span class="p">:</span>
                <span class="n">seqs_dict</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">seqs_dict</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span> <span class="o">=</span> <span class="n">SeqGroup</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">seqs_dict</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">find_common_groups</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">seq_ext_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">],</span> <span class="n">seq_core_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find common sequences between two dictionaries and group them.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_ext_dict: Dictionary of sequences from the first dataset.</span>
<span class="sd">            seq_core_dict: Dictionary of sequences from the second dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dictionary of common sequence mappings and a list of comparison results.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">common</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">group1</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">seq1</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="p">:</span>
                <span class="n">group2</span> <span class="o">=</span> <span class="n">seq_core_dict</span><span class="p">[</span><span class="n">seq1</span><span class="p">]</span>
                <span class="c1"># Check that the 2 groups have the same number of sequences</span>
                <span class="k">if</span> <span class="n">group1</span><span class="o">.</span><span class="n">count</span> <span class="o">==</span> <span class="n">group2</span><span class="o">.</span><span class="n">count</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">group1</span><span class="o">.</span><span class="n">count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">common</span><span class="p">[</span><span class="n">group1</span><span class="o">.</span><span class="n">ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">group2</span><span class="o">.</span><span class="n">ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matched 2 identical groups of sequences: </span><span class="si">{</span><span class="n">group1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">group2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="c1"># Map each ID in group1 to a possible group2 ID</span>
                        <span class="n">possible_id2</span> <span class="o">=</span> <span class="s2">&quot; OR &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">group2</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
                        <span class="n">common</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">id1</span><span class="p">:</span> <span class="n">possible_id2</span> <span class="k">for</span> <span class="n">id1</span> <span class="ow">in</span> <span class="n">group1</span><span class="o">.</span><span class="n">ids</span><span class="p">})</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="s2">&quot;Matched 2 different groups of sequences&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">group1</span><span class="o">.</span><span class="n">count</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">group2</span><span class="o">.</span><span class="n">count</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="n">group1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">group2</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

        <span class="k">return</span> <span class="n">common</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">write_results</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write the comparison results to a file in the output directory.&quot;&quot;&quot;</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;compare.log&quot;</span>
        <span class="n">observed_compare</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Writing results to </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">observed_compare</span><span class="p">:</span>
                    <span class="n">observed_compare</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                    <span class="n">out_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">compare_seq_for_Ns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">only1</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">],</span> <span class="n">only2</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare sequences in `only1` and `only2` for differences in `N` content and length.</span>

<span class="sd">        Args:</span>
<span class="sd">            only1: Sequences unique to the first dataset, mapping sequence to group/identifier.</span>
<span class="sd">            only2: Sequences unique to the second dataset, mapping sequence to group/identifier.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># sequences which have extra N at the end</span>
        <span class="k">for</span> <span class="n">seq_1</span><span class="p">,</span> <span class="n">name1</span> <span class="ow">in</span> <span class="n">only1</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">len1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_1</span><span class="p">)</span>
            <span class="n">seq1_N</span> <span class="o">=</span> <span class="n">seq_1</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">seq_2</span><span class="p">,</span> <span class="n">name2</span> <span class="ow">in</span> <span class="n">only2</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">len2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_2</span><span class="p">)</span>
                <span class="n">seq2_N</span> <span class="o">=</span> <span class="n">seq_2</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>

                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">seq1_N</span> <span class="o">-</span> <span class="n">seq2_N</span><span class="p">)</span> <span class="o">==</span> <span class="nb">abs</span><span class="p">(</span><span class="n">len1</span> <span class="o">-</span> <span class="n">len2</span><span class="p">):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Please check extra Ns added in the accessions </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;ALERT INSERTIONS at the end or diff assembly level </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

                <span class="k">if</span> <span class="n">len1</span> <span class="o">==</span> <span class="n">len2</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">seq2_N</span> <span class="o">&gt;</span> <span class="n">seq1_N</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;your fasta_core has more Ns, check </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sequences have the same length, check </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.CompareFasta.comp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">comp</span> <span class="o">=</span> <span class="p">[]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.CompareFasta.comp" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.CompareFasta.fasta_core" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">fasta_core</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_core</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.CompareFasta.fasta_core" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.CompareFasta.fasta_ext" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">fasta_ext</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_ext</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.CompareFasta.fasta_ext" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.CompareFasta.output_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.CompareFasta.output_dir" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.build_seq_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">build_seq_dict</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.build_seq_dict" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Builds a dictionary of unique sequences and their associated IDs, accounting for duplicates.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>seqs</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where keys are sequence IDs and values are sequences.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;SeqGroup&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.fasta.compare.SeqGroup&lt;/code&gt;)" href="compare/#ensembl.io.genomio.fasta.compare.SeqGroup">SeqGroup</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where keys are unique sequences and values are <code>SeqGroup</code> objects that
group sequence IDs sharing the same sequence.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_seq_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seqs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Builds a dictionary of unique sequences and their associated IDs, accounting for duplicates.</span>

<span class="sd">    Args:</span>
<span class="sd">        seqs: A dictionary where keys are sequence IDs and values are sequences.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dictionary where keys are unique sequences and values are `SeqGroup` objects that</span>
<span class="sd">            group sequence IDs sharing the same sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seqs_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs_dict</span><span class="p">:</span>
            <span class="n">seqs_dict</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">seqs_dict</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span> <span class="o">=</span> <span class="n">SeqGroup</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">seqs_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.compare_seq_for_Ns" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">compare_seq_for_Ns</span><span class="p">(</span><span class="n">only1</span><span class="p">,</span> <span class="n">only2</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.compare_seq_for_Ns" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Compare sequences in <code>only1</code> and <code>only2</code> for differences in <code>N</code> content and length.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>only1</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;SeqGroup&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.fasta.compare.SeqGroup&lt;/code&gt;)" href="compare/#ensembl.io.genomio.fasta.compare.SeqGroup">SeqGroup</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sequences unique to the first dataset, mapping sequence to group/identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only2</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;SeqGroup&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.fasta.compare.SeqGroup&lt;/code&gt;)" href="compare/#ensembl.io.genomio.fasta.compare.SeqGroup">SeqGroup</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sequences unique to the second dataset, mapping sequence to group/identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compare_seq_for_Ns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">only1</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">],</span> <span class="n">only2</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare sequences in `only1` and `only2` for differences in `N` content and length.</span>

<span class="sd">    Args:</span>
<span class="sd">        only1: Sequences unique to the first dataset, mapping sequence to group/identifier.</span>
<span class="sd">        only2: Sequences unique to the second dataset, mapping sequence to group/identifier.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># sequences which have extra N at the end</span>
    <span class="k">for</span> <span class="n">seq_1</span><span class="p">,</span> <span class="n">name1</span> <span class="ow">in</span> <span class="n">only1</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">len1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_1</span><span class="p">)</span>
        <span class="n">seq1_N</span> <span class="o">=</span> <span class="n">seq_1</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">seq_2</span><span class="p">,</span> <span class="n">name2</span> <span class="ow">in</span> <span class="n">only2</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">len2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_2</span><span class="p">)</span>
            <span class="n">seq2_N</span> <span class="o">=</span> <span class="n">seq_2</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">seq1_N</span> <span class="o">-</span> <span class="n">seq2_N</span><span class="p">)</span> <span class="o">==</span> <span class="nb">abs</span><span class="p">(</span><span class="n">len1</span> <span class="o">-</span> <span class="n">len2</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Please check extra Ns added in the accessions </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;ALERT INSERTIONS at the end or diff assembly level </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">len1</span> <span class="o">==</span> <span class="n">len2</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">seq2_N</span> <span class="o">&gt;</span> <span class="n">seq1_N</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;your fasta_core has more Ns, check </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sequences have the same length, check </span><span class="si">{</span><span class="n">name1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.compare_seqs" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">compare_seqs</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.compare_seqs" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Compare two FASTA files for common, unique, and differing sequences.
Use <code>write_results()</code> to generate a report.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compare_seqs</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare two FASTA files for common, unique, and differing sequences.</span>
<span class="sd">    Use `write_results()` to generate a report.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_ext</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_ext</span><span class="p">)</span>
    <span class="n">seq_core</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta_core</span><span class="p">)</span>

    <span class="c1"># Compare sequences</span>
    <span class="n">seq_ext_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_seq_dict</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span>
    <span class="n">seq_core_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_seq_dict</span><span class="p">(</span><span class="n">seq_core</span><span class="p">)</span>

    <span class="c1"># Compare number of sequences</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_core</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;WARNING: Different number of sequences: &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;fasta_ext [ n = </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_ext</span><span class="p">)</span><span class="si">}</span><span class="s2"> ] -Vs- fasta_core [ n = </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_core</span><span class="p">)</span><span class="si">}</span><span class="s2"> ]&quot;</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Different number of sequences: fasta_ext compared to fasta_core&quot;</span><span class="p">)</span>

    <span class="n">common</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">find_common_groups</span><span class="p">(</span><span class="n">seq_ext_dict</span><span class="p">,</span> <span class="n">seq_core_dict</span><span class="p">)</span>

    <span class="c1"># Sequences that are not common</span>
    <span class="n">only1</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq</span><span class="p">:</span> <span class="n">group</span> <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="p">}</span>

    <span class="n">only2</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq</span><span class="p">:</span> <span class="n">group</span> <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="p">}</span>

    <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequences only in Fasta_1: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only1</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sequences only in Fasta_2: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only2</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Common ids: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">common_ids</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">common_ids</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">common</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Check for sequences with extra Ns</span>
    <span class="k">if</span> <span class="n">only1</span> <span class="ow">and</span> <span class="n">only2</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">compare_seq_for_Ns</span><span class="p">(</span><span class="n">only1</span><span class="p">,</span> <span class="n">only2</span><span class="p">)</span>

    <span class="c1"># Write results to file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">write_results</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.find_common_groups" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">find_common_groups</span><span class="p">(</span><span class="n">seq_ext_dict</span><span class="p">,</span> <span class="n">seq_core_dict</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.find_common_groups" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Find common sequences between two dictionaries and group them.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>seq_ext_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;SeqGroup&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.fasta.compare.SeqGroup&lt;/code&gt;)" href="compare/#ensembl.io.genomio.fasta.compare.SeqGroup">SeqGroup</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of sequences from the first dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seq_core_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="&lt;code class=&quot;doc-symbol doc-symbol-heading doc-symbol-class&quot;&gt;&lt;/code&gt;            &lt;code&gt;SeqGroup&lt;/code&gt; (&lt;code&gt;ensembl.io.genomio.fasta.compare.SeqGroup&lt;/code&gt;)" href="compare/#ensembl.io.genomio.fasta.compare.SeqGroup">SeqGroup</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of sequences from the second dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of common sequence mappings and a list of comparison results.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">find_common_groups</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">seq_ext_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">],</span> <span class="n">seq_core_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqGroup</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find common sequences between two dictionaries and group them.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_ext_dict: Dictionary of sequences from the first dataset.</span>
<span class="sd">        seq_core_dict: Dictionary of sequences from the second dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dictionary of common sequence mappings and a list of comparison results.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">common</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">group1</span> <span class="ow">in</span> <span class="n">seq_ext_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">seq1</span> <span class="ow">in</span> <span class="n">seq_core_dict</span><span class="p">:</span>
            <span class="n">group2</span> <span class="o">=</span> <span class="n">seq_core_dict</span><span class="p">[</span><span class="n">seq1</span><span class="p">]</span>
            <span class="c1"># Check that the 2 groups have the same number of sequences</span>
            <span class="k">if</span> <span class="n">group1</span><span class="o">.</span><span class="n">count</span> <span class="o">==</span> <span class="n">group2</span><span class="o">.</span><span class="n">count</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">group1</span><span class="o">.</span><span class="n">count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">common</span><span class="p">[</span><span class="n">group1</span><span class="o">.</span><span class="n">ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">group2</span><span class="o">.</span><span class="n">ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matched 2 identical groups of sequences: </span><span class="si">{</span><span class="n">group1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">group2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="c1"># Map each ID in group1 to a possible group2 ID</span>
                    <span class="n">possible_id2</span> <span class="o">=</span> <span class="s2">&quot; OR &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">group2</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
                    <span class="n">common</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">id1</span><span class="p">:</span> <span class="n">possible_id2</span> <span class="k">for</span> <span class="n">id1</span> <span class="ow">in</span> <span class="n">group1</span><span class="o">.</span><span class="n">ids</span><span class="p">})</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;Matched 2 different groups of sequences&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">group1</span><span class="o">.</span><span class="n">count</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">group2</span><span class="o">.</span><span class="n">count</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="n">group1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">group2</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

    <span class="k">return</span> <span class="n">common</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.read_fasta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">read_fasta</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.read_fasta" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Reads a FASTA file and returns a dictionary mapping sequence IDs to sequences.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>fasta_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the FASTA file. Supports gzipped files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where keys are sequence IDs and values are sequences with all non-CGTA
characters replaced by "N".</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reads a FASTA file and returns a dictionary mapping sequence IDs to sequences.</span>

<span class="sd">    Args:</span>
<span class="sd">        fasta_path: Path to the FASTA file. Supports gzipped files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dictionary where keys are sequence IDs and values are sequences with all non-CGTA</span>
<span class="sd">            characters replaced by &quot;N&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read fasta file </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span>
            <span class="n">sequences</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[^CGTA]&quot;</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
    <span class="k">return</span> <span class="n">sequences</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.CompareFasta.write_results" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">write_results</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.fasta.CompareFasta.write_results" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Write the comparison results to a file in the output directory.</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">write_results</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write the comparison results to a file in the output directory.&quot;&quot;&quot;</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;compare.log&quot;</span>
    <span class="n">observed_compare</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Writing results to </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comp</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">observed_compare</span><span class="p">:</span>
                <span class="n">observed_compare</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="n">out_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.fasta.FastaParserError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>FastaParserError</code>


<a href="#ensembl.io.genomio.fasta.FastaParserError" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="Exception">Exception</span></code></p>


        <p>Error while parsing a FASTA file.</p>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/process.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">35</span>
<span class="normal">36</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">FastaParserError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Error while parsing a FASTA file.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
              </details>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.fasta.SeqGroup" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>SeqGroup</code>


<a href="#ensembl.io.genomio.fasta.SeqGroup" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>Represents a group of sequence identifiers and maintains a count of them.</p>








              <details class="quote">
                <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">SeqGroup</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Represents a group of sequence identifiers and maintains a count of them.&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initializes a `SeqGroup` instance.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifier: The first identifier to add to the group. If `None`, adds &quot;None&quot; as the identifier.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">identifier</span><span class="p">)]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a comma-separated string of sequence identifiers.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a new identifier to the group and updates the count.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifier: The identifier to add. If `None`, &quot;None&quot; is added instead.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">identifier</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.SeqGroup.count" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">count</span> <span class="o">=</span> <span class="mi">1</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.SeqGroup.count" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.io.genomio.fasta.SeqGroup.ids" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">identifier</span><span class="p">)]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.fasta.SeqGroup.ids" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h4 id="ensembl.io.genomio.fasta.SeqGroup.add_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">add_id</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.SeqGroup.add_id" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Add a new identifier to the group and updates the count.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>identifier</code>
            </td>
            <td>
                  <code><span title="str">str</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The identifier to add. If <code>None</code>, "None" is added instead.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/compare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a new identifier to the group and updates the count.</span>

<span class="sd">    Args:</span>
<span class="sd">        identifier: The identifier to add. If `None`, &quot;None&quot; is added instead.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">identifier</span><span class="p">))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.check_chunk_size_and_tolerance" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_chunk_size_and_tolerance</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">,</span> <span class="n">chunk_tolerance</span><span class="p">,</span> <span class="n">error_f</span><span class="o">=</span><span class="n">_on_value_error</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.check_chunk_size_and_tolerance" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Check the chunk size and the tolerance are positive and chunk size is not too small</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Chunk size to check</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_tolerance</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Chunk tolerance to check</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="dies" open>
  <summary>Dies</summary>
  <p>If checks failed dies with <code>parser.error</code></p>
</details>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">check_chunk_size_and_tolerance</span><span class="p">(</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">chunk_tolerance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">error_f</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="n">_on_value_error</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the chunk size and the tolerance are positive and chunk size is not too small</span>

<span class="sd">    Args:</span>
<span class="sd">        chunk_size: Chunk size to check</span>
<span class="sd">        chunk_tolerance: Chunk tolerance to check</span>

<span class="sd">    Dies:</span>
<span class="sd">        If checks failed dies with `parser.error`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">chunk_size</span> <span class="o">&lt;</span> <span class="mi">50_000</span><span class="p">:</span>
        <span class="n">error_f</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong &#39;--chunk_size&#39; value: &#39;</span><span class="si">{</span><span class="n">chunk_size</span><span class="si">}</span><span class="s2">&#39;. should be greater then 50_000. exiting...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">chunk_tolerance</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">error_f</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;wrong &#39;--chunk_tolerance&#39; value: &#39;</span><span class="si">{</span><span class="n">chunk_tolerance</span><span class="si">}</span><span class="s2">&#39;. can&#39;t be less then 0. exiting...&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.chunk_fasta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">chunk_fasta</span><span class="p">(</span><span class="n">input_fasta_file</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">,</span> <span class="n">chunk_size_tolerated</span><span class="p">,</span> <span class="n">out_file_name</span><span class="p">,</span> <span class="n">individual_file_prefix</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">agp_output_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_sequence_len</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">chunk_sfx</span><span class="o">=</span><span class="s1">&#39;ens_chunk&#39;</span><span class="p">,</span> <span class="n">append_offset_to_chunk_name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.chunk_fasta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Open <code>input_fasta_file</code> and split into a smaller chunks based on
stretches of "N"s and then based on chunk_size_tolerated and store either to
the <code>out_file_name</code> if no <code>individual_file_prefix</code> is provided or
store each individual chunk to a file starting with non-empty <code>individual_file_prefix</code>.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>input_fasta_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input FASTA</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of the chunks to split into.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size_tolerated</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If more flexibility allowed, use this as the maximum size of a chunk.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_file_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output FASTA to store the chunks into if no <code>individual_file_prefix</code> is provided.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>individual_file_prefix</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pathlib.Path">Path</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A file path prefix including dirs and filenames part to use as a
    first part of the chunk file name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>agp_output_file</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output AGP file to store the map for the chunking procedure if present and non-empty.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_sequence_len</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of the stretch of <code>N</code>s to split at; not slitting on <code>N</code>s if 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_sfx</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string to put between the original sequence name and the chunk suffix.</p>
              </div>
            </td>
            <td>
                  <code>&#39;ens_chunk&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>append_offset_to_chunk_name</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="bool">bool</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Append 0-based offset in the form of <code>_off_{offset}</code> to the chunk name.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">chunk_fasta</span><span class="p">(</span>
    <span class="n">input_fasta_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">chunk_size_tolerated</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">out_file_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">individual_file_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">],</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">agp_output_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">n_sequence_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">chunk_sfx</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;ens_chunk&quot;</span><span class="p">,</span>
    <span class="n">append_offset_to_chunk_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Open `input_fasta_file` and split into a smaller chunks based on</span>
<span class="sd">    stretches of &quot;N&quot;s and then based on chunk_size_tolerated and store either to</span>
<span class="sd">    the `out_file_name` if no `individual_file_prefix` is provided or</span>
<span class="sd">    store each individual chunk to a file starting with non-empty `individual_file_prefix`.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_fasta_file: Input FASTA</span>
<span class="sd">        chunk_size: Size of the chunks to split into.</span>
<span class="sd">        chunk_size_tolerated: If more flexibility allowed, use this as the maximum size of a chunk.</span>
<span class="sd">        out_file_name: Output FASTA to store the chunks into if no `individual_file_prefix` is provided.</span>
<span class="sd">        individual_file_prefix: A file path prefix including dirs and filenames part to use as a</span>
<span class="sd">                first part of the chunk file name.</span>
<span class="sd">        agp_output_file: Output AGP file to store the map for the chunking procedure if present and non-empty.</span>
<span class="sd">        n_sequence_len: Length of the stretch of `N`s to split at; not slitting on `N`s if 0.</span>
<span class="sd">        chunk_sfx: A string to put between the original sequence name and the chunk suffix.</span>
<span class="sd">        append_offset_to_chunk_name: Append 0-based offset in the form of `_off_{offset}` to the chunk name.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># process input fasta</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">input_fasta_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;splitting sequences from &#39;</span><span class="si">{</span><span class="n">input_fasta_file</span><span class="si">}</span><span class="s2">&#39;, chunk size </span><span class="si">{</span><span class="n">chunk_size</span><span class="si">:</span><span class="s2">_</span><span class="si">}</span><span class="s2">, </span><span class="se">\</span>
<span class="s2">                splitting on </span><span class="si">{</span><span class="n">n_sequence_len</span><span class="si">}</span><span class="s2"> Ns (0 -- disabled)&quot;</span>
        <span class="p">)</span>
        <span class="c1"># do not open a joined file if you plan to open many individual ones</span>
        <span class="k">with</span> <span class="p">(</span>
            <span class="n">individual_file_prefix</span>
            <span class="ow">and</span> <span class="n">nullcontext</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="ow">or</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file_name</span><span class="p">,</span> <span class="s2">&quot;wt&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file_joined</span>
        <span class="p">):</span>
            <span class="n">agp_lines</span> <span class="o">=</span> <span class="n">chunk_fasta_stream</span><span class="p">(</span>
                <span class="n">fasta</span><span class="p">,</span>
                <span class="n">chunk_size</span><span class="p">,</span>
                <span class="n">chunk_size_tolerated</span><span class="p">,</span>
                <span class="n">out_file_joined</span><span class="p">,</span>
                <span class="n">individual_file_prefix</span><span class="p">,</span>
                <span class="n">n_sequence_len</span><span class="o">=</span><span class="n">n_sequence_len</span><span class="p">,</span>
                <span class="n">chunk_sfx</span><span class="o">=</span><span class="n">chunk_sfx</span><span class="p">,</span>
                <span class="n">append_offset_to_chunk_name</span><span class="o">=</span><span class="n">append_offset_to_chunk_name</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># dump AGP</span>
        <span class="k">if</span> <span class="n">agp_output_file</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">agp_output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">agp_out</span><span class="p">:</span>
                <span class="n">agp_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">agp_lines</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.chunk_fasta_stream" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">chunk_fasta_stream</span><span class="p">(</span><span class="n">input_fasta</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">,</span> <span class="n">chunk_size_tolerated</span><span class="p">,</span> <span class="n">output_fasta</span><span class="p">,</span> <span class="n">individual_file_prefix</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">n_sequence_len</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">chunk_sfx</span><span class="o">=</span><span class="s1">&#39;ens_chunk&#39;</span><span class="p">,</span> <span class="n">append_offset_to_chunk_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">open_individual</span><span class="o">=</span><span class="n">_individual_file_opener</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.chunk_fasta_stream" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Split input TextIOWrapper stream with fasta into a smaller chunks based on
stretches of "N"s and then based on chunk_size_tolerated and store either to
the output_fasta stream (if valid) or to the files created by
invocation of the <code>open_individual</code> callable.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>input_fasta</code>
            </td>
            <td>
                  <code><span title="io.TextIOWrapper">TextIOWrapper</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input FASTA as the TextIOWrapper stream.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of the chunks to split into.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size_tolerated</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If more flexibility allowed, use this as the maximum size of a chunk.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_fasta</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="io.TextIOWrapper">TextIOWrapper</span>] | <span title="contextlib.nullcontext">nullcontext</span>[<span title="typing.Any">Any</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output FASTA TextIOWrapper stream to store the chunks into,
    if none or False, <code>open_individual</code> is used (see below).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>individual_file_prefix</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pathlib.Path">Path</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A file path prefix including dirs and filenames part to use as a
    first part of the chunk file name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_sequence_len</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of the stretch of <code>N</code>s to split at; not slitting on <code>N</code>s if 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_sfx</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string to put between the original sequence name and the chunk suffix.</p>
              </div>
            </td>
            <td>
                  <code>&#39;ens_chunk&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>append_offset_to_chunk_name</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="bool">bool</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A flag to append 0-based offset (<code>_off_{offset}</code>) to the chunk name.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>open_individual</code>
            </td>
            <td>
                  <code><span title="typing.Callable">Callable</span>[[<span title="str">str</span>], <span title="typing.ContextManager">ContextManager</span>[<span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A callable taking filename as an input to generate the output file for
    individual contig if out_put FASTA is <code>false</code> of <code>None</code>, folders should be preexisting.</p>
              </div>
            </td>
            <td>
                  <code><span title="ensembl.io.genomio.fasta.chunk._individual_file_opener">_individual_file_opener</span></code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">chunk_fasta_stream</span><span class="p">(</span>
    <span class="n">input_fasta</span><span class="p">:</span> <span class="n">TextIOWrapper</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">chunk_size_tolerated</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">output_fasta</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">TextIOWrapper</span><span class="p">]</span> <span class="o">|</span> <span class="n">nullcontext</span><span class="p">[</span><span class="n">Any</span><span class="p">],</span>
    <span class="n">individual_file_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">],</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">n_sequence_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">chunk_sfx</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;ens_chunk&quot;</span><span class="p">,</span>
    <span class="n">append_offset_to_chunk_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">open_individual</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="n">ContextManager</span><span class="p">[</span><span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="n">_individual_file_opener</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split input TextIOWrapper stream with fasta into a smaller chunks based on</span>
<span class="sd">    stretches of &quot;N&quot;s and then based on chunk_size_tolerated and store either to</span>
<span class="sd">    the output_fasta stream (if valid) or to the files created by</span>
<span class="sd">    invocation of the `open_individual` callable.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_fasta: Input FASTA as the TextIOWrapper stream.</span>
<span class="sd">        chunk_size: Size of the chunks to split into.</span>
<span class="sd">        chunk_size_tolerated: If more flexibility allowed, use this as the maximum size of a chunk.</span>
<span class="sd">        output_fasta: Output FASTA TextIOWrapper stream to store the chunks into,</span>
<span class="sd">                if none or False, `open_individual` is used (see below).</span>
<span class="sd">        individual_file_prefix: A file path prefix including dirs and filenames part to use as a</span>
<span class="sd">                first part of the chunk file name.</span>
<span class="sd">        n_sequence_len: Length of the stretch of `N`s to split at; not slitting on `N`s if 0.</span>
<span class="sd">        chunk_sfx: A string to put between the original sequence name and the chunk suffix.</span>
<span class="sd">        append_offset_to_chunk_name: A flag to append 0-based offset (`_off_{offset}`) to the chunk name.</span>
<span class="sd">        open_individual: A callable taking filename as an input to generate the output file for</span>
<span class="sd">                individual contig if out_put FASTA is `false` of `None`, folders should be preexisting.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chunk_size_tolerated</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">,</span> <span class="n">chunk_size_tolerated</span><span class="p">)</span>
    <span class="c1"># output agp_lines list</span>
    <span class="n">agp_lines</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># make sure not used for n_seq &lt;= 0</span>
    <span class="n">n_split_regex</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">n_sequence_len</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;(N</span><span class="se">{{</span><span class="si">{</span><span class="n">n_sequence_len</span><span class="si">}</span><span class="s2">,</span><span class="se">}}</span><span class="s2">)&quot;</span>
        <span class="n">n_split_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">pattern</span><span class="p">)</span>

    <span class="c1"># process stream</span>
    <span class="n">fasta_parser</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">input_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">rec_count</span><span class="p">,</span> <span class="n">rec</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_parser</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">rec_name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

        <span class="n">ends</span> <span class="o">=</span> <span class="n">split_seq_by_n</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">n_split_regex</span><span class="p">)</span>
        <span class="n">ends</span> <span class="o">=</span> <span class="n">split_seq_by_chunk_size</span><span class="p">(</span><span class="n">ends</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">,</span> <span class="n">chunk_size_tolerated</span><span class="p">)</span>

        <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chunk</span><span class="p">,</span> <span class="n">chunk_end</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ends</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">chunk_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rec_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">chunk_sfx</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">chunk</span><span class="si">:</span><span class="s2">03d</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">chunk_file_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">if</span> <span class="n">individual_file_prefix</span><span class="p">:</span>
                <span class="n">chunk_file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">individual_file_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">rec_count</span><span class="si">:</span><span class="s2">03d</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chunk</span><span class="si">:</span><span class="s2">03d</span><span class="si">}</span><span class="s2">.fa&quot;</span>
            <span class="k">if</span> <span class="n">append_offset_to_chunk_name</span><span class="p">:</span>
                <span class="n">chunk_name</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;_off_</span><span class="si">{</span><span class="n">offset</span><span class="si">}</span><span class="s2">&quot;</span>

            <span class="n">rec_from</span> <span class="o">=</span> <span class="n">offset</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">rec_to</span> <span class="o">=</span> <span class="n">chunk_end</span>
            <span class="n">chunk_len</span> <span class="o">=</span> <span class="n">chunk_end</span> <span class="o">-</span> <span class="n">offset</span>

            <span class="c1"># form agp lines</span>
            <span class="n">agp_line</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rec_name</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">rec_from</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">rec_to</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">chunk</span><span class="si">}</span><span class="se">\t</span><span class="s2">W</span><span class="se">\t</span><span class="si">{</span><span class="n">chunk_name</span><span class="si">}</span><span class="se">\t</span><span class="s2">1</span><span class="se">\t</span><span class="si">{</span><span class="n">chunk_len</span><span class="si">}</span><span class="se">\t</span><span class="s2">+&quot;</span>
            <span class="n">agp_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">agp_line</span><span class="p">)</span>

            <span class="c1"># use agp lines as fasta description</span>
            <span class="n">agp_line</span> <span class="o">=</span> <span class="n">agp_line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dumping </span><span class="si">{</span><span class="n">chunk_name</span><span class="si">}</span><span class="s2"> AGP </span><span class="si">{</span><span class="n">agp_line</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># get slice and put it out</span>
            <span class="n">tmp_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
                <span class="n">Seq</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">offset</span><span class="p">:</span><span class="n">chunk_end</span><span class="p">]),</span>
                <span class="nb">id</span><span class="o">=</span><span class="n">chunk_name</span><span class="p">,</span>
                <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;AGP </span><span class="si">{</span><span class="n">agp_line</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># if user specified chunks -- store each chunk in an individual output file</span>
            <span class="k">with</span> <span class="n">output_fasta</span> <span class="ow">and</span> <span class="n">nullcontext</span><span class="p">(</span><span class="n">output_fasta</span><span class="p">)</span> <span class="ow">or</span> <span class="n">open_individual</span><span class="p">(</span><span class="n">chunk_file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tmp_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;fasta&quot;</span><span class="p">))</span>  <span class="c1"># type: ignore[union-attr]</span>

            <span class="k">del</span> <span class="n">tmp_record</span>
            <span class="n">offset</span> <span class="o">=</span> <span class="n">chunk_end</span>

    <span class="k">return</span> <span class="n">agp_lines</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.get_peptides_to_exclude" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_peptides_to_exclude</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">,</span> <span class="n">seqr_to_exclude</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.get_peptides_to_exclude" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract peptide IDs from a genbank file that are in a given list of seq regions</p>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_peptides_to_exclude</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">seqr_to_exclude</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract peptide IDs from a genbank file that are in a given list of seq regions</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">peptides_to_exclude</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_genbank</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_genbank</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">seqr_to_exclude</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip sequence </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="s2">&quot;protein_id&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                            <span class="n">feat_id</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;protein_id&quot;</span><span class="p">]</span>
                            <span class="n">peptides_to_exclude</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feat_id</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="k">raise</span> <span class="n">FastaParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Peptide without peptide ID $</span><span class="si">{</span><span class="n">feat</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">peptides_to_exclude</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.get_tolerated_size" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_tolerated_size</span><span class="p">(</span><span class="n">size</span><span class="p">,</span> <span class="n">tolerance</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.get_tolerated_size" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate max tolerated size of the chunk based on initial size and percent of allowed deviation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Base chunk size</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tolerance</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Percent of allowed deviance as integer.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum tolerated chunk size.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_tolerated_size</span><span class="p">(</span><span class="n">size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">tolerance</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate max tolerated size of the chunk based on initial size and percent of allowed deviation.</span>

<span class="sd">    Args:</span>
<span class="sd">        size: Base chunk size</span>
<span class="sd">        tolerance: Percent of allowed deviance as integer.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Maximum tolerated chunk size.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tolerance</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">tolerance</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="n">tolerated_size</span> <span class="o">=</span> <span class="n">size</span>
    <span class="n">tolerated_size</span> <span class="o">+=</span> <span class="n">size</span> <span class="o">*</span> <span class="n">tolerance</span> <span class="o">//</span> <span class="mi">100</span>
    <span class="k">return</span> <span class="n">tolerated_size</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.prep_fasta_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">prep_fasta_data</span><span class="p">(</span><span class="n">fasta_infile</span><span class="p">,</span> <span class="n">genbank_infile</span><span class="p">,</span> <span class="n">fasta_outfile</span><span class="p">,</span> <span class="n">peptide_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.prep_fasta_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">



<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>fasta_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input FASTA file - DNA / Protein</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genbank_infile</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input GenBank GBFF file (Optional)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>fasta_outfile</code>
            </td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output FASTA sequence file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>peptide_mode</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Process proteins instead of DNA</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prep_fasta_data</span><span class="p">(</span>
    <span class="n">fasta_infile</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">genbank_infile</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">],</span>
    <span class="n">fasta_outfile</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">peptide_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        fasta_file: Input FASTA file - DNA / Protein</span>
<span class="sd">        genbank_infile: Input GenBank GBFF file (Optional)</span>
<span class="sd">        fasta_outfile: Output FASTA sequence file.</span>
<span class="sd">        peptide_mode: Process proteins instead of DNA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_infile</span><span class="p">)</span>

    <span class="n">to_exclude</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">seqr_to_exclude</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">exclude_seq_regions</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">peptide_mode</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">genbank_infile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">genbank_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genbank_infile</span><span class="p">)</span>
            <span class="n">to_exclude</span> <span class="o">=</span> <span class="n">get_peptides_to_exclude</span><span class="p">(</span><span class="n">genbank_path</span><span class="p">,</span> <span class="n">seqr_to_exclude</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">to_exclude</span> <span class="o">=</span> <span class="n">seqr_to_exclude</span>

    <span class="c1"># Copy and filter</span>
    <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Final path</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_fasta</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip record $</span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">fasta_outfile</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_fasta</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">out_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.prepare_out_dir_for_individuals" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">prepare_out_dir_for_individuals</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">file_part</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.prepare_out_dir_for_individuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates <code>dir_name</code> (including upstream dirs) and returns its paths with the <code>file_part</code> appended.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>dir_name</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to create.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_part</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>File part to append.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pathlib.Path">Path</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><code>dir_name</code> with appended <code>file_part</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="throws" open>
  <summary>Throws</summary>
  <p>exception if not able to create directory.</p>
</details>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prepare_out_dir_for_individuals</span><span class="p">(</span><span class="n">dir_name</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">file_part</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates `dir_name` (including upstream dirs) and returns its paths with the `file_part` appended.</span>

<span class="sd">    Args:</span>
<span class="sd">        dir_name: Directory to create.</span>
<span class="sd">        file_part: File part to append.</span>

<span class="sd">    Returns:</span>
<span class="sd">        `dir_name` with appended `file_part`.</span>

<span class="sd">    Throws:</span>
<span class="sd">        exception if not able to create directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file_prefix</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">dir_name</span><span class="p">:</span>
        <span class="n">dir_name</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">file_prefix</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">file_part</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">file_prefix</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.split_seq_by_chunk_size" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">split_seq_by_chunk_size</span><span class="p">(</span><span class="n">ends</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">,</span> <span class="n">tolerated_size</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.split_seq_by_chunk_size" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Split list of end coordinates, to form chunks not longer then
chunk_size.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>ends</code>
            </td>
            <td>
                  <code><span title="list">list</span>[<span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of one or more chunk(s) to split a sequence.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of chunks to split a sequence into.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tolerated_size</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold to use instead of <code>chunk_size</code> to determine when to split a sequence.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span>[<span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List with open coordinates of the chunks ends (or with only a single sequence length).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">split_seq_by_chunk_size</span><span class="p">(</span>
    <span class="n">ends</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">chunk_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">tolerated_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split list of end coordinates, to form chunks not longer then</span>
<span class="sd">    chunk_size.</span>

<span class="sd">    Args:</span>
<span class="sd">        ends: List of one or more chunk(s) to split a sequence.</span>
<span class="sd">        chunk_size: Size of chunks to split a sequence into.</span>
<span class="sd">        tolerated_size: Threshold to use instead of `chunk_size` to determine when to split a sequence.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List with open coordinates of the chunks ends (or with only a single sequence length).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">ends</span> <span class="ow">or</span> <span class="n">chunk_size</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ends</span>

    <span class="k">if</span> <span class="n">tolerated_size</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">tolerated_size</span> <span class="o">&lt;</span> <span class="n">chunk_size</span><span class="p">:</span>
        <span class="n">tolerated_size</span> <span class="o">=</span> <span class="n">chunk_size</span>
    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">chunk_end</span> <span class="ow">in</span> <span class="n">ends</span><span class="p">:</span>
        <span class="n">chunk_len</span> <span class="o">=</span> <span class="n">chunk_end</span> <span class="o">-</span> <span class="n">offset</span>
        <span class="k">if</span> <span class="n">chunk_len</span> <span class="o">&gt;</span> <span class="n">tolerated_size</span><span class="p">:</span>
            <span class="c1"># exclude starts, as they are 0 or pushed as previous chunk_ends</span>
            <span class="n">result</span> <span class="o">+=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">offset</span><span class="p">,</span> <span class="n">chunk_end</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">))[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chunk_end</span><span class="p">)</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="n">chunk_end</span>
    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.io.genomio.fasta.split_seq_by_n" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">split_seq_by_n</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">split_pattern</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.fasta.split_seq_by_n" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Split a string into chunks at the positions where the
pattern is found. <code>N</code>s (pattern) are appended to the chunk on the left.</p>
<p>The end point of each chunk will correspond to the end
of the matching part.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>seq</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sequence to be split into chunks.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>split_pattern</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="re.Pattern">Pattern</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pattern to search in the sequence.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span>[<span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List with open coordinates of the chunks ends (or with only a single sequence length).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/fasta/chunk.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">split_seq_by_n</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">split_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Pattern</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split a string into chunks at the positions where the</span>
<span class="sd">    pattern is found. `N`s (pattern) are appended to the chunk on the left.</span>

<span class="sd">    The end point of each chunk will correspond to the end</span>
<span class="sd">    of the matching part.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq: Sequence to be split into chunks.</span>
<span class="sd">        split_pattern: Pattern to search in the sequence.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List with open coordinates of the chunks ends (or with only a single sequence length).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">split_pattern</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">seq_len</span><span class="p">]</span>
    <span class="n">split_points</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">split_pattern</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">seq</span><span class="p">)]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">split_points</span> <span class="ow">or</span> <span class="n">split_points</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">seq_len</span><span class="p">:</span>
        <span class="n">split_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">split_points</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../../../..", "features": ["content.tooltips", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../../../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>